### PR TITLE
New event api faceted [Feedback wanted, do not merge now]

### PR DESCRIPTION
**Pull request by exodist**
Just a preview, looking for feedback. Please note, this is a work in progress

@schwern 
@autarch (if you have time)

The basic idea here is that behaviors that are expected of events are encapsulated into 'facets' and an event has facets attached to it.

(An event is restricted to 1 of each type of facet unless otherwise stated)

- trace - not new! but now has a proper home (it felt weird puting it in util:: originally)
- assertion - a facet that is added if an assertion was made
- plan - a facet for if the event makes a plan
- nest - subtest stuff
- Stop - essentially a bail-out
- amnesty - todo skip, anything that makes an assertion that failed not matter (can have multiple)
- Info - note or diag (can have multiple)

There are some additional concepts

- the failure diagnostics are now produced by the TAP formatter for a failed event, they are not diag events, they are generated from the trace at the formatter level
- gravity - This is a display hint, 0 is normal, -1 means do not display, >0 means important info. Essentially a diag is an info facet with >0 gravity, a note is an info facet with 0 gravity
- Most event 'callback' subs are unnecessary now, hub knows what to do now
- Formatter has more info and can make more decisions now, look at the new TAP.pm formatter
- I picked 'emit' for the context method to generate the new events
- All legacy events have been retrofitted to support facets, full back-compat
- I added an optimal 'pass' event, and a `$ctx->pass($name)` shortcut for it, this allows for performance in-line with requirements.

I will try to implement a clone of the Test2::Harness renderer as a formatter for more dogfooding.

### CODE REVIEW

**Comment by exodist** [comment]
^ I have updated the original description int he PR to reflect the latest set of changes

File: lib/Test2/Event/Faceted.pm [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    return 100 if $self->causes_fail;
+    return 0;
+}
+
+sub facets {
+    my $self = shift;
+
+    my %facets = $self->{+_FACETS} ? %{$self->{+_FACETS}} : ();
+
+    push @{$facets{amnesty}} => @{$self->{+_AMNESTY}} if $self->{+_AMNESTY};
```
Comment (autarch) [inline lib/Test2/Event/Faceted.pm]: Won't this (and similar code elsewhere) push onto `$self->{+_FACETS}` over and over every time the `facets` method is called?

Comment (exodist) [inline lib/Test2/Event/Faceted.pm]: Indeed you are correct, I have to fix that

Comment (exodist) [inline lib/Test2/Event/Faceted.pm]: fixed on last push

**Comment by exodist** [comment]
Hmm, I am still not completely happy with this. I think some of the legacy-support logic is too complicated. I also think most of the other methods should be wrapped into facets. The 'nest' stuff also bothers me. Gonna do some more tweaking and see if I can make it cleaner.

**Pull request closed without merging**